ext/uri/php_uriparser.c

```diff
@@ -0,0 +1,429 @@
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->pathHead != NULL) {
+		const UriPathSegmentA *p;
+		smart_str str = {0};
+
+		if (uriparser_uri->absolutePath || uriHasHostA(uriparser_uri)) {
+			smart_str_appendc(&str, '/');
+		}
+
+		smart_str_appendl(&str, uriparser_uri->pathHead->text.first, get_text_range_length(&uriparser_uri->pathHead->text));
+
+		for (p = uriparser_uri->pathHead->next; p != NULL; p = p->next) {
+			smart_str_appendc(&str, '/');
+			smart_str_appendl(&str, p->text.first, get_text_range_length(&p->text));
+		}
+
+		ZVAL_NEW_STR(retval, smart_str_extract(&str));
+	} else if (uriparser_uri->absolutePath) {
+		ZVAL_CHAR(retval, '/');
+	} else {
+		ZVAL_EMPTY_STRING(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_query(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->query.first != NULL && uriparser_uri->query.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->query.first, get_text_range_length(&uriparser_uri->query));
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static zend_result uriparser_read_fragment(const uri_internal_t *internal_uri, uri_component_read_mode_t read_mode, zval *retval)
+{
+	UriUriA *uriparser_uri = uriparser_read_uri(internal_uri->uri, read_mode);
+	ZEND_ASSERT(uriparser_uri != NULL);
+
+	if (uriparser_uri->fragment.first != NULL && uriparser_uri->fragment.afterLast != NULL) {
+		ZVAL_STRINGL(retval, uriparser_uri->fragment.first, get_text_range_length(&uriparser_uri->fragment));
+	} else {
+		ZVAL_NULL(retval);
+	}
+
+	return SUCCESS;
+}
+
+static void *uriparser_malloc(UriMemoryManager *memory_manager, size_t size)
+{
+	return emalloc(size);
+}
+
+static void *uriparser_calloc(UriMemoryManager *memory_manager, size_t nmemb, size_t size)
+{
+	return ecalloc(nmemb, size);
+}
+
+static void *uriparser_realloc(UriMemoryManager *memory_manager, void *ptr, size_t size)
+{
+	return erealloc(ptr, size);
+}
+
+static void *uriparser_reallocarray(UriMemoryManager *memory_manager, void *ptr, size_t nmemb, size_t size)
+{
+	return safe_erealloc(ptr, nmemb, size, 0);
+}
+
+static void uriparser_free(UriMemoryManager *memory_manager, void *ptr)
+{
+	efree(ptr);
+}
+
+PHP_MINIT_FUNCTION(uri_uriparser)
+{
+	if (uri_handler_register(&uriparser_uri_handler) == FAILURE) {
+		return FAILURE;
+	}
+
+	defaultMemoryManager.malloc = uriparser_malloc;
+	defaultMemoryManager.calloc = uriparser_calloc;
+	defaultMemoryManager.realloc = uriparser_realloc;
+	defaultMemoryManager.reallocarray = uriparser_reallocarray;
+	defaultMemoryManager.free = uriparser_free;
+
+	return SUCCESS;
+}
+
+static uriparser_uris_t *uriparser_create_uris(void)
+{
+	uriparser_uris_t *uriparser_uris = emalloc(sizeof(*uriparser_uris));
```

Comment:
```suggestion
	uriparser_uris_t *uriparser_uris = ecalloc(1, sizeof(*uriparser_uris));
```
This would cleanly zero-initialize everything. As far as I can see, the uriparser library guarantees that calling `uriFreeUriMembersA()` on a zero-initialized URI is also safe, thus this improves the overall safety somewhat.

Reply:
Asked if this guarantee will continue to hold: https://github.com/uriparser/uriparser/issues/238